src/compiler/checker.ts

```diff
@@ -13349,23 +13349,60 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return result;
     }
 
+    function getFallbackExpandedLabels(restType: TupleTypeReference, restDeclaration: Declaration | undefined) {
+        if (!restDeclaration) {
+            return;
+        }
+        Debug.assertNode(restDeclaration, isParameter);
+        if (restDeclaration.name.kind !== SyntaxKind.ArrayBindingPattern) {
+            return;
+        }
+        const nameElements = restDeclaration.name.elements;
+        const elementFlags = restType.target.elementFlags;
+        const result: (__String | undefined)[] = arrayOf(elementFlags.length, () => undefined);
+        const length = Math.min(nameElements.length, elementFlags.length);
+        for (let i = 0; i < length; i++) {
+            const nameElement = nameElements[i];
+
+            if (elementFlags[i] & ElementFlags.Variable) {
+                if (nameElement.kind === SyntaxKind.BindingElement && nameElement.dotDotDotToken && nameElement.name.kind === SyntaxKind.Identifier) {
+                    result[i] = nameElement.name.escapedText;
+                }
+                break;
+            }
+            if (nameElement.kind !== SyntaxKind.BindingElement) {
+                continue;
+            }
+            if (nameElement.dotDotDotToken) {
+                break;
+            }
+            if (nameElement.name.kind !== SyntaxKind.Identifier) {
+                continue;
+            }
+            result[i] = nameElement.name.escapedText;
+        }
+        return result;
+    }
+
     function getExpandedParameters(sig: Signature, skipUnionExpanding?: boolean): readonly (readonly Symbol[])[] {
         if (signatureHasRestParameter(sig)) {
             const restIndex = sig.parameters.length - 1;
             const restName = sig.parameters[restIndex].escapedName;
             const restType = getTypeOfSymbol(sig.parameters[restIndex]);
+            const restDeclaration = sig.parameters[restIndex].valueDeclaration;
+
             if (isTupleType(restType)) {
-                return [expandSignatureParametersWithTupleMembers(restType, restIndex, restName)];
+                return [expandSignatureParametersWithTupleMembers(restType, restIndex, restName, getFallbackExpandedLabels(restType, restDeclaration))];

```

Comment:
Another thing I realized is that we're going to be calculating this fallback list every time we do this, even though it's not always used. But, I'm not sure how to make that lazy either.

Reply:
ye, this is somewhat unfortunate but I don't think it's a particularly costly operation to create this so I just don't mind having this eager here